src/nnetbin/nnet-train-multistream.cc

```diff
@@ -0,0 +1,433 @@
+                num_other_error++;
+                continue;
+              }
+
+              // get the (feature,target) pair,
+              Matrix<BaseFloat> mat = feature_reader.Value();
+              Posterior targets = target_reader.Value(utt);
+
+              // getting per-frame weihts,
+              Vector<BaseFloat> weights;
+              if (frame_weights != "") {
+                weights = weights_reader.Value(utt);
+              } else {  // all per-frame weights are 1.0
+                weights.Resize(mat.NumRows());
+                weights.Set(1.0);
+              }
+
+              // correct small length mismatch ... or drop sentence
+              {
+                // add lengths to vector
+                std::vector<int32> length;
+                length.push_back(mat.NumRows());
+                length.push_back(targets.size());
+                length.push_back(weights.Dim());
+                // find min, max
+                int32 min = *std::min_element(length.begin(), length.end());
+                int32 max = *std::max_element(length.begin(), length.end());
+                // fix or drop ?
+                if (max - min < length_tolerance) {
+                  if (mat.NumRows() != min) mat.Resize(min, mat.NumCols(), kCopyData);
+                  if (targets.size() != min) targets.resize(min);
+                  if (weights.Dim() != min) weights.Resize(min, kCopyData);
+                } else {
+                  KALDI_WARN << "Length mismatch! Targets " << targets.size()
+                             << ", features " << mat.NumRows() << ", " << utt;
+                  num_other_error++;
+                  continue;
+                }
+              }
+
+              // input transform may contain splicing,
+              nnet_transf.Feedforward(CuMatrix<BaseFloat>(mat), &feats_transf);
+
+              /* Here we could do the 'targets_delay', BUT...
+               * It is better to do it by a <Splice> component!
+               *
+               * The prototype would look like this (6th frame becomes 1st frame, etc.):
+               * '<Splice> <InputDim> dim1 <OutputDim> dim1 <BuildVector> 5 </BuildVector>'
+               */
+
+              // store,
+              feats_utt[s] = Matrix<BaseFloat>(feats_transf);
+              labels_utt[s] = targets;
+              weights_utt[s] = weights;
+              new_utt_flags[s] = 1;
+
+              // end the loop,
+              feature_reader.Next();
+              break;
+            }
+          }
+        }  // for 'num_streams'
+      }  // Re-fill,
+
+      // end the training after processing all the frames,
+      size_t frames_to_go = 0;
+      for (int32 s = 0; s < num_streams; s++) {
+        frames_to_go += feats_utt[s].NumRows();
+      }
+      if (frames_to_go == 0) break;
+
+      // number of frames we'll pack as the streams,
+      std::vector<int32> frame_num_utt;
+
+      // pack the parallel data,
```

Comment:
It seems to me like the code would be easier to follow if you broke some things out into functions.
This is way past the "soft limit" of 40 lines per function.